marimo/_ai/_tools/tools/tables_and_variables.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2025 Marimo. All rights reserved.
+from dataclasses import dataclass, field
+from typing import Any, Optional
+
+from marimo._ai._tools.base import ToolBase
+from marimo._ai._tools.tools.cells import CellVariableValue
+from marimo._ai._tools.types import SuccessResult
+from marimo._server.sessions import Session
+from marimo._types.ids import SessionId
+
+
+@dataclass
+class TablesAndVariablesArgs:
+    session_id: SessionId
+    variable_names: list[str]
+
+
+@dataclass
+class ColumnInfo:
+    name: str
+    data_type: str
+    external_type: str
+    sample_values: list[Any]
+
+
+@dataclass
+class DataTableMetadata:
+    source: str
+    num_rows: int
+    num_columns: int
+    columns: list[ColumnInfo]
+    primary_keys: Optional[list[str]]
+    indexes: Optional[list[str]]
+
+
+@dataclass
+class TablesAndVariablesOutput(SuccessResult):
+    tables: dict[str, DataTableMetadata] = field(default_factory=dict)
+    variables: dict[str, CellVariableValue] = field(default_factory=dict)
+
+
+class GetTablesAndVariables(
+    ToolBase[TablesAndVariablesArgs, TablesAndVariablesOutput]
+):
+    """
+    Get tables and variables information in the session.
+
+    When provided with a list of variable names, it will return information about the variables and tables mentioned.
+    If an empty list is provided, it will return information about all tables and variables.
+
+    Returns:
+        A success result containing tables (columns, primary keys, indexes, etc.) and variables (value, data type).
+    """
+
+    def handle(self, args: TablesAndVariablesArgs) -> TablesAndVariablesOutput:
+        session_manager = self.context.session_manager
+        session = session_manager.get_session(args.session_id)
```

Comment:
you should use `get_session` on the `self.context` as it provides a better error message